cvmfs/file_watcher_inotify.cc

```diff
@@ -0,0 +1,167 @@
+
+namespace file_watcher {
+
+FileWatcherInotify::FileWatcherInotify() {}
+
+FileWatcherInotify::~FileWatcherInotify() {}
+
+bool FileWatcherInotify::RunEventLoop(const FileWatcher::HandlerMap& handlers,
+                                      int read_pipe, int write_pipe) {
+  inotify_fd_ = inotify_init1(IN_NONBLOCK);
+  assert(inotify_fd_ >= 0);
+
+  for (FileWatcher::HandlerMap::const_iterator it = handlers.begin();
+       it != handlers.end(); ++it) {
+    RegisterFilter(it->first, it->second);
+  }
+
+  // Use the control pipe to signal readiness to the main thread,
+  // before continuing with the event loop.
+  WritePipe(write_pipe, "s", 1);
+
+  struct pollfd poll_set[2];
+
+  poll_set[0].fd = read_pipe;
+  poll_set[0].events = POLLHUP | POLLIN;
+  poll_set[0].revents = 0;
+  poll_set[1].fd = inotify_fd_;
+  poll_set[1].events = POLLIN;
+  poll_set[1].revents = 0;
+
+  bool stop = false;
+  while (!stop) {
+    int ready = poll(poll_set, 2, -1);
+    if (ready == -1) {
+      LogCvmfs(kLogCvmfs, kLogSyslogErr,
+               "FileWatcherInotify - Could not poll events. Errno: %d", errno);
+      return false;
+    }
+    if (ready == 0) {
+      continue;
+    }
+
+    if (poll_set[0].revents & POLLHUP) {
+      LogCvmfs(kLogCvmfs, kLogDebug, "FileWatcherInotify - Stopping.\n");
+      stop = true;
+      continue;
+    }
+    if (poll_set[0].revents & POLLIN) {
+      char buffer[1];
+      ReadPipe(read_pipe, &buffer, 1);
+      LogCvmfs(kLogCvmfs, kLogDebug, "FileWatcherInotify - Stopping.\n");
+      stop = true;
+      continue;
+    }
+
+    const size_t event_size = sizeof(struct inotify_event);
+    const size_t buffer_size = event_size + 100;
+    char buffer[buffer_size];
+    if (poll_set[1].revents & POLLIN) {
+      int len = read(inotify_fd_, buffer, buffer_size);
+      assert(len > 0);
+      int i = 0;
+      while (i < len) {
+        struct inotify_event* inotify_event =
+            reinterpret_cast<struct inotify_event*>(&buffer[i]);
+        std::map<int, WatchRecord>::const_iterator it =
+            watch_records_.find(inotify_event->wd);
+        if (it != watch_records_.end()) {
+          WatchRecord current_record = it->second;
+          file_watcher::Event event = file_watcher::kInvalid;
+          if (inotify_event->mask & IN_DELETE_SELF) {
+            event = file_watcher::kDeleted;
+          } else if (inotify_event->mask & IN_MODIFY) {
+            // Modified
+            event = file_watcher::kModified;
+          } else if (inotify_event->mask & IN_MOVE_SELF) {
+            // Renamed
+            event = file_watcher::kRenamed;
+          } else if (inotify_event->mask & IN_ATTRIB) {
+            // Attributes
+            event = file_watcher::kAttributes;
+          } else if (inotify_event->mask & IN_IGNORED) {
+            // Attributes
+            event = file_watcher::kIgnored;
+          }
+          bool clear_handler = true;
+          if (event != file_watcher::kInvalid ||
```

Comment:
Shouldn't this be "event not invalid _and_ event not ignored"?

Reply:
Good catch.